<a href="https://colab.research.google.com/github/AburizalAN/learn-ml/blob/master/database/07_database_project_Aburizal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Project: Database


**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
2. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive if you haven't already. This will be the repository for all your completed assignment and project files, aiding you in keeping your work organized and accessible.
3. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
4. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
5. **Setting Permission toPublic**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.


## Student Identity


In [1]:
# @title #### Student Identity
student_id = "REAEAYH8" # @param {type:"string"}
name = "Aburizal Adi Nugroho" # @param {type:"string"}
drive_link = "https://drive.google.com/drive/folders/1pRIbA2Zphrtg_TS5N0ypHHTXkUBs50tI?usp=sharing"  # @param {type:"string"}

assignment_id = "00_database_project"

# Import grader package
!pip install rggrader
from rggrader import submit, submit_image

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.


Rmember to make a copy of this notebook in your Google Drive and work in your own copy.


## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.


### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.


In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below



### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.


In [2]:
# Write your data loading to SQL database here
import sqlite3
import pandas as pd

dataset = pd.read_csv("./data.csv", delimiter=",")
conn = sqlite3.connect('ecommerce.db')
cur = conn.cursor()

# cur.execute('''
#   CREATE TABLE ecommerce (
#     id INT PRIMARY KEY,
#     InvoiceNo VARCHAR(255) NOT NULL,
#     StockCode VARCHAR(255) NOT NULL,
#     Description TEXT NOT NULL,
#     Quantity INT NOT NULL,
#     InvoiceDate DATETIME NOT NULL,
#     UnitPrice FLOAT NOT NULL,
#     CustomerID FLOAT NOT NULL,
#     Country VARCHAR(255) NOT NULL
#   )
# ''')

dataset.to_sql('ecommerce', conn, if_exists="replace", index=False)

conn.commit()

In [3]:
import pandas as pd
pd.read_sql_query("select * from ecommerce where StockCode = '85123A'", conn)

InvoiceNo StockCode                         Description  Quantity  \
0       536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1       536373    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
2       536375    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
3       536390    85123A  WHITE HANGING HEART T-LIGHT HOLDER        64   
4       536394    85123A  WHITE HANGING HEART T-LIGHT HOLDER        32   
...        ...       ...                                 ...       ...   
2308    581439    85123A  CREAM HANGING HEART T-LIGHT HOLDER         1   
2309    581452    85123A  WHITE HANGING HEART T-LIGHT HOLDER        32   
2310    581472    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
2311    581492    85123A  CREAM HANGING HEART T-LIGHT HOLDER         3   
2312    581538    85123A  CREAM HANGING HEART T-LIGHT HOLDER         1   

          InvoiceDate  UnitPrice  CustomerID         Country  
0      12/1/2010 8:26       2.55     17850.0  United Kingdom  
1      12/1/2010 9:02       2.55     17850.0  United Kingdom  
2      12/1/2010 9:32       2.55     17850.0  United Kingdom  
3     12/1/2010 10:19       2.55     17511.0  United Kingdom  
4     12/1/2010 10:39       2.55     13408.0  United Kingdom  
...               ...        ...         ...             ...  
2308  12/8/2011 16:30       5.79         NaN  United Kingdom  
2309  12/8/2011 18:03       2.55     17675.0  United Kingdom  
2310  12/8/2011 19:55       2.95     15796.0  United Kingdom  
2311  12/9/2011 10:03       5.79         NaN  United Kingdom  
2312  12/9/2011 11:34       2.95     14446.0  United Kingdom  

[2313 rows x 8 columns]

### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909


In [4]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('ecommerce.db') #change db_name to your database name.
cur = conn.cursor()

# Get column names
cur.execute("PRAGMA table_info(ecommerce)") #change table_name to your table name
sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
print(f"Columns: {sql_columns}")

# Get total number of rows
cur.execute("SELECT COUNT(*) FROM ecommerce") #change table_name to your table name
sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {sql_num_rows}")

# Close the connection
conn.close()

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

5. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption


In [5]:
# Write your SQL Query here

#use the following variable name to store the result from your SQL Query
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below

conn = sqlite3.connect('ecommerce.db') #change db_name to your database name.
cur = conn.cursor()

cur.execute("select COUNT(DISTINCT StockCode) from ecommerce")
sql_num_products = cur.fetchone()[0]

cur.execute("select COUNT(DISTINCT CustomerID) from ecommerce")
sql_num_customers = cur.fetchone()[0]

conn.close()

sql_num_products


# customerIDTable = pd.read_sql_query("select CustomerID, COUNT(*) as 'CustomerID_Count' from ecommerce GROUP BY CustomerID", conn)

# 1. Find the total number of unique products and unique customers
# Example: sql_num_products = 1559
# Example: sql_num_customers = 1675




4070

In [6]:
# 2. Calculate the total revenue for each product, limited to top 5
# Example: sql_top_product_revenues = [('PRODUCT 1', 149385.12), ('PRODUCT 2', 54895.53), ('PRODUCT 3', 44545.55), ('PRODUCT 4', 38000.00), ('PRODUCT 5', 36000.00)]
# cur.execute("select StockCode, SUM(Quantity) from ecommerce group by StockCode")
conn = sqlite3.connect('ecommerce.db') #change db_name to your database name.
cur = conn.cursor()

cur.execute("""
  select
    Description,
    sum(Quantity * UnitPrice)
  from ecommerce
    GROUP BY Description
    ORDER BY sum(Quantity * UnitPrice) DESC
    LIMIT 5
""")

sql_top_product_revenues = cur.fetchall()
conn.close()

sql_top_product_revenues



[('DOTCOM POSTAGE', 206245.48000000019),
 ('REGENCY CAKESTAND 3 TIER', 164762.19),
 ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849),
 ('PARTY BUNTING', 98302.98000000048),
 ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]

In [7]:
# 3. Identify the most profitable countries (by total sales), limited to top 5
# Example: sql_top_countries = [('COUNTRY 1', 600000.00), ('COUNTRY 2', 334857.45), ('COUNTRY 3', 245879.00), ('COUNTRY 4', 180000.00), ('COUNTRY 5', 164389.45)]

conn = sqlite3.connect('ecommerce.db') #change db_name to your database name.
cur = conn.cursor()

cur.execute("""
  select
    Country,
    SUM(Quantity * UnitPrice)
  from ecommerce
    GROUP BY Country
    ORDER BY SUM(Quantity * UnitPrice) DESC
    LIMIT 5
""")

sql_top_countries = cur.fetchall()

conn.close()

sql_top_countries


[('United Kingdom', 8187806.364001113),
 ('Netherlands', 284661.54000000015),
 ('EIRE', 263276.81999999826),
 ('Germany', 221698.20999999862),
 ('France', 197403.90000000002)]

In [8]:
# 4. Find the top-selling products (by total sales) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 200000.00), ('COUNTRY 1 PRODUCT 2', 100000.00), ('COUNTRY 1 PRODUCT 3', 95348.00), ('COUNTRY 1 PRODUCT 4', 90000.43), ('COUNTRY 1 PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000.00), ('COUNTRY 2 PRODUCT 2', 7345.00), ('COUNTRY 2 PRODUCT 3', 6934.23), ('COUNTRY 2 PRODUCT 4', 6800.23), ('COUNTRY 2 PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 7442.84), ('COUNTRY 5 PRODUCT 2', 5175.00), ('COUNTRY 5 PRODUCT 3', 3098.0), ('COUNTRY 5 PRODUCT 4', 3014.34), ('COUNTRY 5 PRODUCT 5', 2857.83)],
import sqlite3
import pandas as pd

conn = sqlite3.connect('ecommerce.db') #change db_name to your database name.
cur = conn.cursor()

sql_top_selling_products_by_country_by_sales = {}

for i in range(len(sql_top_countries)):
  cur.execute('''
    select Description, SUM(Quantity * UnitPrice)
    from ecommerce
      where Country = ?
      group by Description
      order by SUM(Quantity * UnitPrice) DESC
      limit 5
  ''', [sql_top_countries[i][0]])
  data = cur.fetchall()
  sql_top_selling_products_by_country_by_sales[sql_top_countries[i][0]] = data

conn.close()

print(sql_top_selling_products_by_country_by_sales)

{'United Kingdom': [('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 134405.93999999992), ('WHITE HANGING HEART T-LIGHT HOLDER', 93953.0699999987), ('PARTY BUNTING', 92501.73000000033), ('JUMBO BAG RED RETROSPOT', 84516.44000000166)], 'Netherlands': [('RABBIT NIGHT LIGHT', 9568.48), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 7991.400000000001), ('SPACEBOY LUNCH BOX ', 7485.5999999999985), ('DOLLY GIRL LUNCH BOX', 6828.599999999999), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 4039.2000000000007)], 'EIRE': [('REGENCY CAKESTAND 3 TIER', 7442.849999999997), ('CARRIAGE', 5175.0), ('JAM MAKING SET WITH JARS', 3089.0), ('3 TIER CAKE TIN RED AND CREAM', 3041.5499999999997), ('WHITE HANGING HEART T-LIGHT HOLDER', 2857.8)], 'Germany': [('POSTAGE', 20821.0), ('REGENCY CAKESTAND 3 TIER', 8257.349999999997), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3554.6999999999985), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 1949.9500000000016), ('SPACEBOY LUNCH BOX ', 1629.45)], 'France': [('POSTAGE', 1

In [9]:
# 5. Find the top-selling products (by quantity) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_quantity = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 22343), ('COUNTRY 1 PRODUCT 2', 12345), ('COUNTRY 1 PRODUCT 3', 9534), ('COUNTRY 1 PRODUCT 4', 9000), ('COUNTRY 1 PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000), ('COUNTRY 2 PRODUCT 2', 7345), ('COUNTRY 2 PRODUCT 3', 6934), ('COUNTRY 2 PRODUCT 4', 6800), ('COUNTRY 2 PRODUCT 5', 4000)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 48345), ('COUNTRY 5 PRODUCT 2', 43965), ('COUNTRY 5 PRODUCT 3', 43556), ('COUNTRY 5 PRODUCT 4', 34567), ('COUNTRY 5 PRODUCT 5', 33436)],
import sqlite3
import pandas as pd

conn = sqlite3.connect('ecommerce.db') #change db_name to your database name.
cur = conn.cursor()

sql_top_selling_products_by_country_by_quantity = {}

for i in range(len(sql_top_countries)):
  cur.execute('''
    select Description, SUM(Quantity)
    from ecommerce
      where Country = ?
      group by Description
      order by SUM(Quantity) DESC
      limit 5
  ''', [sql_top_countries[i][0]])
  data = cur.fetchall()
  sql_top_selling_products_by_country_by_quantity[sql_top_countries[i][0]] = data

conn.close()

print(sql_top_selling_products_by_country_by_quantity)

{'United Kingdom': [('WORLD WAR 2 GLIDERS ASSTD DESIGNS', 48326), ('JUMBO BAG RED RETROSPOT', 43167), ('POPCORN HOLDER', 34365), ('ASSORTED COLOUR BIRD ORNAMENT', 33679), ('WHITE HANGING HEART T-LIGHT HOLDER', 33193)], 'Netherlands': [('RABBIT NIGHT LIGHT', 4801), ('SPACEBOY LUNCH BOX ', 4528), ('DOLLY GIRL LUNCH BOX', 4132), ('PACK OF 72 RETROSPOT CAKE CASES', 4128), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)], 'EIRE': [('PACK OF 72 RETROSPOT CAKE CASES', 1728), ('60 TEATIME FAIRY CAKE CASES', 1536), ('VINTAGE SNAP CARDS', 1492), ('ASSORTED INCENSE PACK', 1440), ('PACK OF 60 PINK PAISLEY CAKE CASES', 1296)], 'Germany': [('ROUND SNACK BOXES SET OF4 WOODLAND ', 1218), ('ASSORTED COLOURS SILK FAN', 1164), ('POSTAGE', 1104), ('WOODLAND CHARLOTTE BAG', 1019), ('PACK OF 72 RETROSPOT CAKE CASES', 1002)], 'France': [('RABBIT NIGHT LIGHT', 4023), ('MINI PAINT SET VINTAGE ', 2196), ('RED TOADSTOOL LED NIGHT LIGHT', 1314), ('SET/6 RED SPOTTY PAPER CUPS', 1272), ('ASSORTED COLOUR BIRD ORNAMENT

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.


In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below



### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.


In [12]:
# Preprocessing here
import pandas as pd
from elasticsearch import Elasticsearch, helpers
import json
import time


es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])
df = pd.read_csv('./data.csv')
df.dropna(inplace=True)

def df_to_es(df):
  for index, row in df.iterrows():
    yield {
      "_index": "ecommerce",
      "_id": index,
      "_source": row.to_dict()
    }

res = helpers.bulk(es, df_to_es(df))

print(res)


# Load your dataset and create an Elasticsearch index
# Hints: use the BulkAPI for faster creation of the index



(406829, [])


### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829


In [13]:
from elasticsearch import Elasticsearch

# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = "ecommerce" #change the my_index_name value

# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 541909


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.

2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.

3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.

By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.

There are two more tasks below to check your understanding on how to handle date type data as well as searching using partial keyword:

4. **Find a product's transactions over time**

   Trends in purchases over time for specific products can provide insights into product popularity and seasonality. You decide to track a product's transactions over a period to understand its sales pattern. In this case we'll use the product "I LOVE LONDON MINI RUCKSACK"

5. **Search for a product using a part of its description**

   Finally, you want to investigate how robust and reliable your product search feature is. You decide to test it by searching for a product using a part of its description. In this case, we'll use the keyword "PIN". Return only unique product name. And the product names should be stored in a list in alphabetical order.


In [14]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below
es_transactions_over_time = {} #dictionary where key is invoice date, value is the quantity sold, see example below
es_unique_product_search = [] #list containing the products in alphabetical order

# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584

response = es.search(index='ecommerce', body={
  'query': {
    'term': {
      'Country.keyword': 'Germany'
    }
  },
})

print(json.dumps(response.body, indent=4))
es_transactions_in_germany = response['hits']['total']['value']

{
    "took": 11,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 9495,
            "relation": "eq"
        },
        "max_score": 4.0342965,
        "hits": [
            {
                "_index": "ecommerce",
                "_id": "1109",
                "_score": 4.0342965,
                "_source": {
                    "InvoiceNo": "536527",
                    "StockCode": "22809",
                    "Description": "SET OF 6 T-LIGHTS SANTA",
                    "Quantity": 6,
                    "InvoiceDate": "12/1/2010 13:04",
                    "UnitPrice": 2.95,
                    "CustomerID": 12662.0,
                    "Country": "Germany"
                }
            },
            {
                "_index": "ecommerce",
                "_id": "1110",
                "_score": 4.0342965,
                "_source": {
     

In [ ]:
# 2. Find all unique products in Germany
# Hints: Use set() to remove duplicate values
# Example: es_total_unique_products = 45345




In [ ]:
# 3. Find the top 5 most purchased products in Germany
# Example: es_top_products = {'PRODUCT 1': 498, 'PRODUCT 2': 452, 'PRODUCT 3': 342, 'PRODUCT 4': 231, 'PRODUCT 5': 123}




In [ ]:
# 4. Find a product's transactions over time
# Hints: Date format is M/d/yyyy H:m
# Product = I LOVE LONDON MINI RUCKSACK
# Example: es_transactions_over_time = {'12/12/2010 11:11': 1, 12/12/2010 11:12': 1}




In [ ]:
# 5. Search for a product using a part of its description
# Hints: Return only unique Product names, see number 1
# Example: es_unique_product_search = ['PRODUCT 1', 'PRODUCT 2']




## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [16]:
# Submit Method

# Do not change the code below
# question_id = "01_sql_columns"
# submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
# question_id = "02_sql_num_rows"
# submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
# question_id = "03_sql_num_products"
# submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
# question_id = "04_sql_num_customers"
# submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
# question_id = "05_sql_top_product_revenues"
# submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
# question_id = "06_sql_top_countries"
# submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
# question_id = "07_sql_top_selling_products_by_country_by_sales"
# submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
# question_id = "08_sql_top_selling_products_by_country_by_quantity"
# submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

print(es_columns)
print(es_num_rows)

question_id = "09_es_columns"
submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
question_id = "10_es_num_rows"
submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
question_id = "11_es_total_unique_products"
submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
# question_id = "12_es_transactions_in_germany"
# submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
# question_id = "13_es_top_products"
# submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)
# question_id = "14_es_transactions_over_time"
# submit(student_id, name, assignment_id, str(es_transactions_over_time), question_id, drive_link)
# question_id = "15_es_unique_product_search"
# submit(student_id, name, assignment_id, str(es_unique_product_search), question_id, drive_link)

['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
541909


'Assignment successfully submitted'

## FIN
